hs/src/Reach/AST/SL.hs

```diff
@@ -160,6 +160,127 @@ data SLVal
+instance (Equiv a) => Equiv (Maybe a) where
+  equiv a b = case (a, b) of
+    (Nothing, Nothing) -> True
+    (Just a', Just b') -> equiv a' b'
+    _ -> False
+
+instance (Equiv a) => Equiv (S.Set a) where
+  equiv s1 s2 = equiv (S.toList s1) (S.toList s2)
+
+instance Equiv B.ByteString where
+  equiv = (==)
+
+instance Equiv Char where
+  equiv = (==)
+
+instance Equiv Int where
+  equiv = (==)
+
+instance Equiv Integer where
+  equiv = (==)
+
+instance Equiv Bool where
+  equiv = (==)
+
+instance Equiv SLTypeFun where
+  equiv (SLTypeFun {stf_pre = pre1, stf_post = post1}) (SLTypeFun {stf_pre = pre2, stf_post = post2}) =
+    equiv pre1 pre2 && equiv post1 post2
+
+instance Equiv SLSSVal where
+  equiv (SLSSVal {sss_val = v1}) (SLSSVal { sss_val = v2}) = equiv v1 v2
+
+instance Equiv DLMVar where
+  equiv (DLMVar a) (DLMVar b) = equiv a b
+
+instance (Equiv a) => Equiv [a] where
+  equiv xs ys = all (\(x,y) -> equiv x y) $ zip xs ys
+
+instance (Equiv b, Equiv a) => Equiv (M.Map a b) where
+  equiv m1 m2 = equiv (M.toList m1) (M.toList m2)
+
+instance (Equiv a, Equiv b) => Equiv (a, b) where
+  equiv (a,b) (a2,b2) = equiv a a2 && equiv b b2
+
+
+instance Equiv SLForm where
+  equiv a b = case (a,b) of
+    (SLForm_App, SLForm_App) -> True
+    (SLForm_each, SLForm_each) -> True
+    ((SLForm_EachAns parts _ _ _), (SLForm_EachAns parts2 _ _ _)) -> equiv parts parts2
+    ((SLForm_Part_Only p _), (SLForm_Part_Only p2 _)) -> equiv p p2
+    ((SLForm_liftInteract p _ _), (SLForm_liftInteract p2 _ _)) -> equiv p p2
+    ((SLForm_Part_ToConsensus {slptc_whos = who, slptc_mv = mv}), (SLForm_Part_ToConsensus {slptc_whos = who2, slptc_mv = mv2})) ->
+      equiv who who2 && equiv mv mv2
+    (SLForm_unknowable, SLForm_unknowable) -> True
+    (SLForm_fork, SLForm_fork) -> True
+    (SLForm_parallel_reduce, SLForm_parallel_reduce) -> True
+    (SLForm_wait, SLForm_wait) -> True
+    -- these partials are left unchecked
+    (SLForm_parallel_reduce_partial{}, _) -> False
+    (SLForm_fork_partial{}, _) -> False
+    _ -> False
+
+instance Equiv SLType where
+  equiv a b = case (a,b) of
+    (ST_Null, ST_Null) -> True
+    (ST_Bool, ST_Bool) -> True
+    (ST_UInt, ST_UInt) -> True
+    (ST_Digest, ST_Digest) -> True
+    (ST_Address, ST_Address) -> True
+    (ST_Token, ST_Token) -> True
+    ((ST_Bytes i), (ST_Bytes i2)) -> equiv i i2
+    ((ST_Array s1 _len1), (ST_Array s2 _len2)) -> equiv s1 s2
+    ((ST_Tuple xs), (ST_Tuple ys)) -> equiv xs ys
+    ((ST_Object m), (ST_Object m2)) -> equiv m m2
+    ((ST_Data m), (ST_Data m2)) -> equiv m m2
+    ((ST_Struct s), (ST_Struct s2)) -> equiv s s2
+    ((ST_Fun f), (ST_Fun f2)) -> equiv f f2
+    ((ST_UDFun f), (ST_UDFun f2)) -> equiv f f2
+    ((ST_Type s1), (ST_Type s2)) -> equiv s1 s2
+    ((ST_Refine _sLType v _m), (ST_Refine _sLType2 v2 _m2)) -> equiv v v2
+    _ -> False
+
+instance Equiv DLVar where
+  equiv (DLVar _ sl _dl i) (DLVar _ sl2 _dl2 i2) =
+    equiv (fmap snd sl) (fmap snd sl2) && i == i2
```

Comment:
We actually only care about the numbers. Also, I like to have everything have a suffix if anything does, so I'd do `i1` and `i2` or maybe `x` and `y`